infoschema/builder.go

```diff
@@ -127,6 +129,30 @@ func (b *Builder) applyCreateSchema(m *meta.Meta, diff *model.SchemaDiff) error
 	return nil
 }
 
+func (b *Builder) applyModifySchemaCharsetAndCollate(m *meta.Meta, diff *model.SchemaDiff) error {
+	di, err := m.GetDatabase(diff.SchemaID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if di == nil {
+		// When we apply an old schema diff, the database may has been dropped already
+		return ErrDatabaseNotExists.GenWithStackByArgs(
+			fmt.Sprintf("(Schema ID %d)", diff.SchemaID),
+		)
+	}
+	oldInfo, ok := b.is.SchemaByName(di.Name)
+	if !ok {
+		return ErrDatabaseNotExists.GenWithStackByArgs(di.Name.O)
+	}
+	if oldInfo.Charset == di.Charset && oldInfo.Collate == di.Collate {
```

Comment:
If this happens, it is already processed when the DDL job is processed. Doesn't this happen here?

Reply:
I suppose another DDL job may have changed the charset already, is it possible?

Reply:
I don't think so. If another DDL job changed, this DDL job needn't to be handled.